lib/ansible/inventory/group.py

```diff
@@ -96,6 +130,17 @@ def add_child_group(self, group):
 
         # don't add if it's already there
         if group not in self.child_groups:
+
+            # prepare list of group's new ancestors this edge creates
+            start_ancestors = group.get_ancestors()
+            new_ancestors = self.get_ancestors()
+            new_ancestors.add(self)
+            new_ancestors.difference_update(start_ancestors)
+            if group in new_ancestors:
+                raise AnsibleError(
+                    "Adding group '%s' as child to '%s' creates a recursive "
+                    "dependency loop." % (group.name, self.name))
```

Comment:
It would be more ideal to move this check above the 1 or 2 lines above that modify `new_ancestors`. Reason is in the test assertion for `get_ancestors`. If the network _already_ contains recursion then the method will return `self`  in the set.

That means that `start_ancestors ` could contain `group`, which would subtract it from `new_ancestors`, so we can avoid that by just doing the check before the subtraction.